src/lsm/tree.zig

```diff
@@ -898,10 +898,18 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
                         context.compaction.reset();
                         tree.manifest.remove_invisible_tables(
                             context.compaction.level_b,
-                            context.compaction.snapshot,
+                            tree.prefetch_snapshot_max,
                             context.compaction.range.key_min,
                             context.compaction.range.key_max,
                         );
+                        if (context.compaction.level_b > 0) {
+                            tree.manifest.remove_invisible_tables(
+                                context.compaction.level_b - 1,
+                                tree.prefetch_snapshot_max,
+                                context.compaction.range.key_min,
+                                context.compaction.range.key_max,
+                            );
+                        }
```

Comment:
It looks like the bug here was even as simple as that we were never removing the table that was compacted from level A?

Reply:
i.e.

1. We compact input tables from level A and level B.
2. We expect to remove all these input tables after compaction.
3. We were only removing the input tables from level B.

Reply:
That was part of the bug. The other part is that we were using the wrong snapshot to determine whether the table was invisible at all.

I think the original idea is that we would only remove from level B, and tables in level A would be removed in the next half-measure (when they were level B tables). But that only works if you remove the key range (which would be much slower for large manifest levels).

Reply:
Nice fix.